### ISSUE TITLE
implement cluster validation using indices

### ISSUE THREAD

**Issue by sreichl** [epigen/unsupervised_analysis#10]
- [x] structure this into coherent notes & subtasks
- [x] implement external clustering indices
- [x] visualize external indices
- [x] implement internal clustering indices
- [x] #21 
- [x] visualize internal indices
- [x] investigate if there is a bug in metadata conversion to factor in validation_internal.R
- [x] create issue(s) for not implemented ideas
- [x] test on Lee2020NatGenet
- [x] test robsutness of cluster indices in different sample proportions 0.1, 0.25, 0.5, 0.75, 0.9, 1 on digits

**internal cluster indices**
- Question to answer: Which clustering separates/groups "objectively" the data the best?
- input
  - original data/distance matrix (which metric?!)
  - all clustering results
  - categorical metadata
- output:
  - cluster_validation/internal_indices_clustering.csv/.png
  - cluster_validation/internal_indices_metadata.csv/.png
  - ?internal_indices_ALL.png (indices & metadata combined)
  - **consider** one plot consisting of a panel of three (differently sized) heatmaps
 - pseudocode
  1. determine scores for clusterings and categorical metadata (requires distance matrix)
  2. determine rank using MCDM/MCDA e.g., TOPSIS
     MCDM / MCDA python package with TOPSIS: https://pymcdm.readthedocs.io/en/master/index.html
       - [x] check if better alternatives exist
  3. save ranked score matrix as CSV
  4. visualize as heatmap w/ & w/o metadata (scale by scores)
    - clusterings (and/or metadata) as rows are ordered by MCDM ranking
    - indices as columns are hierarchically clustered
- indices
  - internal clustering indices: Silhouette, Calinski-Harabasz, Tau, C-index (used in master thesis)
  - use Dunn index instead of Tau index
  - add Davis-Bouldin Score
  - Akaike Information Criterion (self-made in Master Thesis)
  - Bayesian Information Criteria (self-made in Master Thesis)
    - https://towardsdatascience.com/are-you-still-using-the-elbow-method-5d271b3063bd
  - @sreichl [Master Thesis](https://catalogplus.tuwien.at/primo-explore/fulldisplay?docid=UTW_alma2180892620003336&context=L&adaptor=Local%20Search%20Engine&vid=UTW&lang=de_DE&search_scope=UTW&tab=default_tab&query=addsrcrid,exact,AC15010381) Chapter 4.2.2. Internal Indices

**external cluster indices**
- Question to answer: Which metadata explains the clustering best?
  - Benchmark scenario: Which clustering most resembles the ground truth?
- input
  - all clustering results
  - categorical metadata
- output:
  - cluster_validation/external_indices_{index}.csv
  - cluster_validation/external_indices.png
- pseudocode
  1. determine scores for clusterings vs categorical metadata
  3. save the score matrix as CSV
  4. visualize each index as heatmap -> panel of heatmaps
    - clusterings as rows are hierarchically clustered
    - categorical metadata as columns are hierarchically clustered
- indices
  - ARI & NMI
    ```python
    # metrics
    from sklearn.metrics.cluster import adjusted_rand_score
    from sklearn.metrics.cluster import normalized_mutual_info_score
    
    # external cluster indices with ground truth
    adjusted_rand_score(labels, partition.membership)
    normalized_mutual_info_score(labels, partition.membership)
    ```

![image](https://github.com/epigen/unsupervised_analysis/assets/53785552/6352a4f9-2c2c-45d4-81ad-0e21328482e6)

**Comment by sreichl** [comment]
internal clustering indices (self-made in Master Thesis)
- Akaike Information Criterion
- Bayesian Information Criteria
    - https://towardsdatascience.com/are-you-still-using-the-elbow-method-5d271b3063bd

**Comment by sreichl** [comment]
```python
# metrics
from sklearn.metrics.cluster import adjusted_rand_score
from sklearn.metrics.cluster import normalized_mutual_info_score

# external cluster indices with ground truth
adjusted_rand_score(labels, partition.membership)
normalized_mutual_info_score(labels, partition.membership)
```

**Comment by sreichl** [comment]
Significance analysis for clustering with single-cell RNA-sequencing data
https://www.nature.com/articles/s41592-023-01933-9

**Comment by sreichl** [comment]
MCDM / MCDA python package with TOPSIS: https://pymcdm.readthedocs.io/en/master/index.html
- [ ] check if better alternatives exist

**Comment by sreichl** [comment]
fast distance matrix computation
- https://www.kaggle.com/code/cpmpml/ultra-fast-distance-matrix-computation
- https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.euclidean_distances.html

**Comment by sreichl** [comment]
implemented external indices (because in sklearn available)
- [AMI](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_mutual_info_score.html#sklearn.metrics.adjusted_mutual_info_score)
- [ARI](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_rand_score.html#sklearn.metrics.adjusted_rand_score)
- [FMI](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.fowlkes_mallows_score.html#sklearn.metrics.fowlkes_mallows_score)
- [**Homogeneity** and **Completeness** and **V**-Measure](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.homogeneity_completeness_v_measure.html#sklearn.metrics.homogeneity_completeness_v_measure)

**Comment by sreichl** [comment]
decisions

- distance fixed euclidean? -> yes, bc no easy implementation around
- sklearn with 3 or clusterCrit with many more? -> clusterCriot, bc performance is the same but more options and code
- TOPSIS in R or python? -> python bc R package does not exist anymore (and not on conda)
- Information Criteria approach? -> yes it adds something unique
- configurable criteria? -> no

**Comment by sreichl** [comment]
Plot internal indices 

- scale columns
- Cluster only columns 
- Show all at the same time as large heatmap (metadata & clusterings)
- Topsis ranked clusterings and metadata (rows)
- Consider separate script for visualization

**Comment by sreichl** [comment]
implemented internal indices from clusterCrit package:

- Silhouette
- Calinski-Harabasz
- C-index 
- Dunn index
- Davis-Bouldin Score
- Akaike Information Criterion (self-made in Master Thesis)
- Bayesian Information Criteria (self-made in Master Thesis)

https://rdrr.io/cran/clusterCrit/man/intCriteria.html
https://rdrr.io/cran/clusterCrit/f/inst/doc/clusterCrit.pdf

**Comment by sreichl** [comment]
MCDM/A with TOPSIS using [pymcdm](https://pymcdm.readthedocs.io/en/master/example/Example.html)

**Comment by sreichl** [comment]
internal cluster index calculation does not scale to large data sets.
ideas to improve performance of clusterCrit:

increase expression in R ?
options(expressions=10000)

increase memory to e.g., 100G?

-> if data is large determine indices on the top PCs? -> how many? eg 90% variance??

ALL clusterings and 1 score in top 10 PCs: Time difference of 2.379581 mins

1 clustering and 1 score in top 10PCs: Time difference of 5.583407 secs

-> factor 25.57 and all clusterings are 26 so linear scaling O(n) -> no action

1 clustering and 5 scores: Time difference of 1.873907 mins

-> factor 20 -> more than linear, keep scores split up.

effect of more features: 1 clustering and 1 score in top 20PCs: Time difference of 8.466122 secs

-> factor 1.5 although features 2x

effect of observations: 2x observations 1 clustering 1 score top 10 PCs: Time difference of 22.43458 secs

-> factor 4 although observations factor 2x -> O(2n)


TODO: test effect of using PCA with x% e.g., 0.9 of variation vs original data in internal cluster indices in small data sets
ALTERNATIVE/combination: sample 10% of data to determine scores (with random_state to reproduce)?
-> 100% of data and 10PCs: Sil=0.1542688
-> 10% of data and 10PCs: Sil=0.1434478 -> Time difference of 0.06139326 secs
-> 10% of data and 10PCs: Sil=0.1545943
-> 10% of data and 10PCs: Sil=0.1461169
-> 10% of data and 90% of PCs: Sil=-0.05937944

data 2166x9064 with 1 clustering and 1 score -> 3.736456 minutes (factor 4000, although features are factor 900) -> TOTAL time would be 3.7*26->96m=1.5h

**Comment by sreichl** [comment]
decision: remove AIC (arguments below)

- clearly prefers more clusters over less and thereby biases the MCDM approach, which is equal weighted across scores
- correlates highly with BIC (which penalizes more clusters/variables more)
